text/2022-11-08-font-compiler-ir.md

```diff
@@ -0,0 +1,190 @@
+_Why_ is articulated in `TODO: link to dfrg's doc when posted`
+
+- [Goals](#goals)
+  - [Potential future goals](#potential-future-goals)
+  - [Non-goals](#non-goals)
+- [Basic structure](#basic-structure)
+- [IR](#ir)
+  - [Font sources](#font-sources)
+  - [Sketch](#sketch)
+  - [Change tracking](#change-tracking)
+  - [Implementation notes](#implementation-notes)
+  - [Test inputs](#test-inputs)
+- [References](#references)
+
+# Goals
+
+Provide a common format all current and anticipated font source formats can be converted into by a frontend so we can have a single backend compiler to produce
+binary fonts.
+
+The backend should be variable-first (no [merge](https://simoncozens.github.io/compiling-variable-fonts/#merge)). Once the user has a variable font they can
+instance and subset it down as they see fit.
+
+The design should facilitate parallel execution (see
+[Simon’s explorations](https://github.com/googlefonts/oxidize/blob/main/text/2022-05-10-parallel-font-compile-experiments.md#what-discrete-operations-are-available))
+and incremental compilation.
+
+We should be able to optionally capture IR. IR should support human friendly (textual) and fast/compact (binary) representation (shamelessly copying ... I mean
+inspired by ... [LLVM IR](https://llvm.org/docs/Reference.html#llvm-ir), hopefully leaning on serde).
+
+## Potential future goals
+
+Enable translation from `sourceA → IR → sourceB` so we can modernise sources in an automated manner (credit @cmyr for this idea).
+
+## Non-goals
+
+Lossless `source → IR → source`.
+
+# Basic structure
+
+```
+Source → IR → font binary
+            ^ one ir2font implementation
+       ^ many source2ir implementations
+```
+
+Source is one of
+
+- .glyphs
+- UFO (including many w/design space)
+- .ttx
+- nanoemoji config
+
+In future, sources could additionally include:
+
+- binary font
+  - Some contend if we support ttx we effectively support binary fonts
+- Unknown Future Design Format
+
+For every source we will have a dedicated conversion to IR, e.g. `glyphs2ir`, `ufo2ir`, etc in https://github.com/googlefonts/fontmake-rs.
+
+Let’s take `ufo2ir` as a concrete example, it would:
+
+- Depend on a UFO parser, likely https://github.com/linebender/norad
+  - Aside: norad uses quick-xml, we should try to stick to that elsewhere
+- Depend on the IR library
+
+# IR
+
+## Font sources
+
+Rough outline of common source structures
+
+UFO:
+
+- Designspace file maps UFO(s) into design space
+- UFO font has layers
+- Layers have glyphs
```

Comment:
interesting, how does that work concretely? Does the UFO default layer's contents.plist file define the font's glyphs set and then the presence of given named glyph in the other non-default layers count as 'that glyph having that layer' (rather than the other way around)?
And how does this way of thinking of UFOs interact with designspace? I suppose you rely on mutliple DS source elements all pointing to the same UFO filename and having a `layer` attribute set to select where to source glyph definitions.

Reply:
> the presence of given named glyph in the other non-default layers count as 'that glyph having that layer' 

Yes, that is the easy bit.

I'm not claiming it is trivial to do perfect round-tripping, and my implementation isn't even complete. It can still deal with multiple UFOs and multiple layers per UFO, and I maintain a mapping between "glyph layer identity" and "ufo source + layerName". It will become somewhat more complex once Fontra is able to create new glyph layers and sources that don't already exist in the DS system as it is.

Reply:
I'm realizing that I don't really understand what "layer" means in this context or what it maps to in the final binary?

Reply:
Both in .glyphs and UFO/designspace, a glyph layer contains the master shape for a specific designspace location. Additionnally, such layers are used by designers as a scratch layer or a background layer, which will be ignored for the final binary.

Reply:
For a bare UFO, only the default layer matters to a compiler. For a Designspace, every default layer of every referenced UFO matters, plus the DS sources that reference a specific layer inside a UFO. All other layers have no meaning to the compiler.

Reply:
Do you happen to have an example of a font using a DS source that references a specific layer lying around?

Reply:
https://github.com/LettError/mutatorSans/blob/master/MutatorSans.designspace